File: kanz/pkg/bus/retry.go
package bus

import (
	"context"
	"time"
)

// RetryConfig bounds in-handler retry. MaxAttempts is the TOTAL attempts
// including the first one; 0 or 1 means "no retry, surface or DLQ on first
// failure". InitialBackoff doubles each attempt up to MaxBackoff.
//
// Retry runs inside the Consumer's Subscribe dispatch — it blocks the
// underlying broker delivery for the duration of the retry sequence, so
// long backoffs or large MaxAttempts trade per-partition throughput for
// resilience.
type RetryConfig struct {
	MaxAttempts    int
	InitialBackoff time.Duration
	MaxBackoff     time.Duration
}

const (
	defaultMaxAttempts    = 1
	defaultInitialBackoff = 100 * time.Millisecond
	defaultMaxBackoff     = 30 * time.Second
)

func (c RetryConfig) withDefaults() RetryConfig {
	if c.MaxAttempts < 1 {
		c.MaxAttempts = defaultMaxAttempts
	}
	if c.InitialBackoff <= 0 {
		c.InitialBackoff = defaultInitialBackoff
	}
	if c.MaxBackoff <= 0 {
		c.MaxBackoff = defaultMaxBackoff
	}
	return c
}

// Backoff returns the delay before the (attempt+1)-th try. `attempt` is
// 1-based — the first failure passes attempt=1 and gets InitialBackoff.
// Exponential with cap: initial, 2·initial, 4·initial, …, capped at
// MaxBackoff. No jitter at this layer; operational tuning is a follow-up.
func (c RetryConfig) Backoff(attempt int) time.Duration {
	if attempt < 1 {
		return 0
	}
	c = c.withDefaults()
	d := c.InitialBackoff
	for i := 1; i < attempt; i++ {
		d *= 2
		if d >= c.MaxBackoff || d < 0 { // <0 guards Duration overflow
			return c.MaxBackoff
		}
	}
	return d
}

func sleepWithCtx(ctx context.Context, d time.Duration) error {
	if d <= 0 {
		return nil
	}
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-t.C:
		return nil
	}
}

// The DLQ subject convention, header contract and terminal-error signal moved
// to dlq.go when #220 gave the DLQ a drain path: they are the wire contract
// between the side that PARKS a message and the side that DRAINS one, and that
// is a different concept from the in-handler retry budget this file bounds.

File: kanz-py/kanz_bus/retry.py
"""Bounded retry config + DLQ wire conventions (EVT-18d).

Mirrors the Go ``RetryConfig`` (EVT-17e). Retry runs inside the Consumer's
dispatch — it blocks the broker delivery for the duration of the sequence,
so long backoffs × large ``max_attempts`` trade per-partition throughput
for resilience.
"""

from __future__ import annotations

from dataclasses import dataclass

_DEFAULT_MAX_ATTEMPTS = 1
_DEFAULT_INITIAL_BACKOFF = 0.1
_DEFAULT_MAX_BACKOFF = 30.0


@dataclass
class RetryConfig:
    """Total attempts including the first; 0 or 1 means "no retry, surface
    or DLQ on first failure". Backoff doubles each attempt up to
    ``max_backoff_seconds``. No jitter at this layer (operational hardening
    is a follow-up)."""

    max_attempts: int = _DEFAULT_MAX_ATTEMPTS
    initial_backoff_seconds: float = _DEFAULT_INITIAL_BACKOFF
    max_backoff_seconds: float = _DEFAULT_MAX_BACKOFF

    def backoff(self, attempt: int) -> float:
        """Seconds to wait before the (attempt+1)-th try; ``attempt`` is 1-based."""
        if attempt < 1:
            return 0.0
        init = self.initial_backoff_seconds if self.initial_backoff_seconds > 0 else _DEFAULT_INITIAL_BACKOFF
        cap = self.max_backoff_seconds if self.max_backoff_seconds > 0 else _DEFAULT_MAX_BACKOFF
        d = init
        for _ in range(1, attempt):
            d *= 2
            if d >= cap:
                return cap
        return d


# DLQ subject convention: ``dlq.<original-subject>``. Matches the NATS
# ``dlq.>`` stream and per-topic ``dlq.{name}`` Kafka topics already
# provisioned in EVT-08/09.
DLQ_SUBJECT_PREFIX = "dlq."
